### ISSUE TITLE
Add ability to have upload file/media as meta type

### ISSUE THREAD

**Issue by AaronHolbrook** [mboynes/super-cpt#4]
Using new media uploader - would be nice to have the ability to upload files and attach them to the post.

**Comment by mboynes** [comment]
I've considered this. http://wordpress.org/extend/plugins/multiple-post-thumbnails/ does such a great job, I didn't know if it was worth reinventing the wheel or not. I'll re-evaluate

**Comment by AaronHolbrook** [comment]
Yea I'm using that for now, thought it would be good though for things where you don't need just a featured thumbnail, but say a PDF.

I vote yes :)

**Comment by scottlee** [comment]
I too would love to see this! It's the one thing keeping me from switching entirely from WPAlchemy. Thanks for this sweet plugin Matthew!

**Comment by mboynes** [comment]
Related idea from [@GreggFranklin](https://twitter.com/greggfranklin) on Twitter: Add support for including the featured image in custom columns

**Comment by AaronHolbrook** [comment]
Not sure I follow - custom columns? Also - it wouldn't need to be a featured image, just the ability to have file uploads.

**Comment by mboynes** [comment]
In admin lists, e.g. using the the manage_${post_type}_posts_columns filter. You can do this per-field by adding `'column' => true` to the field declaration. His idea was adding the ability to do this for the featured image, so the lists in admin would look something like the media page.

**Comment by AaronHolbrook** [comment]
Ahhh - hmm, that doesn't make as much sense to me, seems like the realm of a separate plugin, at least a separate ticket.

**Comment by mboynes** [comment]
The reason why I added it here is, if/when I add the media "type", it will need to handle the column option (which is already a part of the plugin and works for the other field types). If the media is an image, I'd want to handle it with a thumb. So long as I'm doing that, I may as well add a line or two of code to add support for core featured images, too. It _seems like_ a separate issue, but if you were in my head, you'd see that this is just a scribble on the spec.

**Comment by mosheeshel** [comment]
Anything new? when is the next release planned?

**Comment by mboynes** [comment]
Initial commit on this: 8727f97b70919ea04c8af94ab9662908556f804d

**Comment by Gyroscopic** [comment]
Hi! Love the plugin :) 
I've integrated it into a theme and also added extra cpt fields for colour and time which use the inbuilt wp colorpicker and the jquery timepicker addon. 

Anyway, I was wondering how you are getting on with adding the Featured Image column to the list pages (as well as the multiple-post-thumbnails support - which I've also used on another project and is awesome!) as I will be using this?

**Comment by N-Molham** [comment]
I have a media uploader repo here which could be implemented with Super-CPT
https://github.com/N-Molham/wp-media-uploader

**Comment by mboynes** [comment]
The current master branch has multiple post thumbnail support; I'm waiting for the release of 3.6 to deploy it, however, to see if the MP6 admin updates make it in. With the update, I did add column support for the additional thumbnails, but I haven't yet added support for the featured images. I probably will before the next release.

**Comment by Gyroscopic** [comment]
Hi Matt

Thanks for all the updates. I'm trying to work out how to implement the media columns for featured images. Currently there is a featured image meta box (on the side) displaying when creating/editing the cpt, which I presume is because of the default post type settings for the 'supports' field having thumbnail set. However, the featured image is not showing in the cpt listing page as a column - how do I set this column to be displayed?

If I add a 'media' meta field in a metabox and set the columns to true then this displays as a column on the listing page but I then have 2 separate image fields and inputs which I dont want.

I am using the latest version as of today.

**Comment by mboynes** [comment]
You can display the featured image column by calling `$cpt->add_to_columns( '_thumbnail_id' );` This method will display any post meta, whether SuperCPT created it or not. I'll be updating the docs this week with information about this and other new features.

**Comment by Gyroscopic** [comment]
Hi Matt thanks for the quick reply.

I tried that but it only displays the image id in the column. I've also tried '_wp_attached_file' and '_wp_attachment_metadata' which seem to be other meta-keys but nothing works. I cant seem to find any code in the plugin that uses the equivalent of...

```
$featured_image = wp_get_attachment_image_src( get_post_thumbnail_id( $post->ID ), 'thumbnail' );
```

To retrieve and display the image in the column - am I missing something?

**Comment by mboynes** [comment]
Can you verify that you're using the latest version? [Here is the relevant code in the source](https://github.com/mboynes/super-cpt/blob/master/includes/class-super-custom-post-meta.php#L798-L799). Here is an example of what your code should look like:

```
$demo_posts = new Super_Custom_Post_Type( 'demo-post' );
$demo_posts->add_to_columns( '_thumbnail_id' );
```

**Comment by Gyroscopic** [comment]
Hi just checked and am using the latest version. 

Upon investigation I added a var_dump($data); to the format_meta_for_list function in class-super-custom-post-meta.php and I get **two outputs for each media item!** The first is the id and second is the file path/name. 
Code...

``` php
(we set $xxx=0 at the top of the class)
    public function format_meta_for_list( $data, $key ) {
        global $xxx;
        $field_info = get_known_field_info( $key, $this->type );
        if ( false == $field_info ) {
            switch ( $key ) {
                case '_thumbnail_id' :
                        var_dump($data);
                    if (++$xxx & 1 ) return wp_get_attachment_image( $data, $this->column_thumbnail_size, true );
            }
```

Output...
![cap](https://f.cloud.github.com/assets/4580445/930746/c12ce114-0014-11e3-9985-9debc25cd240.PNG)

Not sure why it calls the function twice for each post, but I added a counter and check if its odd to only return the file path/name and it works to display the thumbnail.

Let me know if its a bug or not.

As a side note it would be great if we could also change the column title when we add thumbnail columns as each cpt is different and therefore the thumbnail has a different context.
Possibly, $cpt->add_to_columns( '_thumbnail_id', 'Title'); ?

btw if you want my code for the colour picker and time picker, let me know.
